homeassistant/components/kostal_piko/__init__.py

```diff
@@ -0,0 +1,118 @@
+"""Kostal Piko custom component."""
+from datetime import timedelta
+import logging
+from math import ceil
+
+from kostal import InfoVersions, Piko, SettingsGeneral
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_HOST, CONF_PASSWORD, CONF_USERNAME
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers.aiohttp_client import async_get_clientsession
+from homeassistant.helpers.update_coordinator import DataUpdateCoordinator, UpdateFailed
+
+from .const import DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+PLATFORMS = ["sensor"]
+DEVICE_INFO_IDS = [
+    SettingsGeneral.INVERTER_NAME,
+    SettingsGeneral.INVERTER_MAKE,
+    InfoVersions.SERIAL_NUMBER,
+    InfoVersions.ARTICLE_NUMBER,
+    InfoVersions.VERSION_FW,
+    InfoVersions.VERSION_HW,
+]
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Kostal Piko from a config entry."""
+    host = entry.data[CONF_HOST]
+    user = entry.data[CONF_USERNAME]
+    password = entry.data[CONF_PASSWORD]
+
+    try:
+        piko = Piko(async_get_clientsession(hass), host, user, password)
+    except Exception as err:
+        raise ConfigEntryNotReady from err
+
+    coordinator = PikoUpdateCoordinator(hass, piko)
+    await coordinator.async_config_entry_first_refresh()
+
+    hass.data.setdefault(DOMAIN, {})
+    hass.data[DOMAIN][entry.entry_id] = coordinator
```

Comment:
You can combine these

```suggestion
    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = coordinator
```